File: ProgrammingWCFServices/CPP/SimpleWCF/SimpleWCFService/CServiceHostWrapper.cpp
#include "StdAfx.h"
#include "CServiceHostWrapper.h"
//-------------------------------------------------------------------------------------------------------
namespace SimpleWCFService
{
//-------------------------------------------------------------------------------------------------------
generic<typename T>
CServiceHostWrapper<T>::CServiceHostWrapper(void)
{
	MSyn=gcnew Object();
	MHost=gcnew ServiceHost(T::typeid);

	MHost->OpenTimeout=TimeSpan(0,0,10);
	MHost->CloseTimeout=TimeSpan(0,0,10);
}
//-------------------------------------------------------------------------------------------------------
generic<typename T>
CServiceHostWrapper<T>::~CServiceHostWrapper(void)
{
	StopService();
}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
generic<typename T>
ServiceHost^ CServiceHostWrapper<T>::Host::get(void)
{
	return(MHost);
}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
generic<typename T>
generic<typename TL>
void CServiceHostWrapper<T>::AddServiceEndpoint(Binding^ EndpointBinding, String^ Address)
{
	lock			LockRoot(MSyn);
	{
		try
		{
			MHost->AddServiceEndpoint(TL::typeid,EndpointBinding,Address);
		}
		catch(Exception^ E)
		{
			Console::WriteLine(String::Format("CServiceHostWrapper.AddServiceEndpoint() - Exception: {0} !",E->Message));
		}
	}
}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
generic<typename T>
void CServiceHostWrapper<T>::StartService(void)
{
	lock			LockRoot(MSyn);
	{
		try
		{
			MHost->Open();
		}
		catch(Exception^ E)
		{
			Console::WriteLine(String::Format("CServiceHostWrapper.StartService() - Exception: {0} !",E->Message));
		}
	}
}
//-------------------------------------------------------------------------------------------------------
generic<typename T>
void CServiceHostWrapper<T>::StopService(void)
{
	lock			MLockRoot(MSyn);
	{
		try
		{
			if (MHost->State==CommunicationState::Opened==true || MHost->State==CommunicationState::Opening==true)
				MHost->Close();
		}
		catch(Exception^ E)
		{
			Console::WriteLine(String::Format("CServiceHostWrapper.StopService() - Exception: {0} !",E->Message));
		}
	}
}
//-------------------------------------------------------------------------------------------------------
generic<typename T>
void CServiceHostWrapper<T>::AddTCPMEXEndpoint(String^ MEXEndpoindName)
{
	lock			MLockRoot(MSyn);
	{
		try
		{
			BindingElement^				BE=gcnew TcpTransportBindingElement();
			CustomBinding^				CB=gcnew CustomBinding(BE);
			ServiceMetadataBehavior^	SMB;

			if ((SMB=Host->Description->Behaviors->Find<ServiceMetadataBehavior^>())==nullptr)
			{
				SMB=gcnew ServiceMetadataBehavior();
				Host->Description->Behaviors->Add(SMB);
			}

			MHost->AddServiceEndpoint(IMetadataExchange::typeid,CB,MEXEndpoindName);
		}
		catch(Exception^ E)
		{
			Console::WriteLine(String::Format("CServiceHostWrapper.AddServiceEndpoint() - Exception: {0} !",E->Message));
		}
	}
}
//-------------------------------------------------------------------------------------------------------
}
//-------------------------------------------------------------------------------------------------------

File: ProgrammingWCFServices/Chapter09/ResponseMSMQ/ResponseCalculatorService/CServiceHostWrapper.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.ServiceModel;
using System.ServiceModel.Channels;
using System.ServiceModel.Description;
//-------------------------------------------------------------------------------------------------------
namespace ResponseCalculatorService
{
//-------------------------------------------------------------------------------------------------------
	public class CServiceHostWrapper<T> : IServiceHostWrapper
	{
//-------------------------------------------------------------------------------------------------------
		private object											MSyn;
		private ServiceHost										MHost;
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
		public CServiceHostWrapper()
		{
			MSyn=new object();
			MHost=new ServiceHost(typeof(T));

			MHost.OpenTimeout=new TimeSpan(0,0,10);
			MHost.CloseTimeout=new TimeSpan(0,0,10);
		}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
		public ServiceHost										Host
		{
			get
			{
				return(MHost);
			}
		}
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
//-------------------------------------------------------------------------------------------------------
		public void StartService()
		{
			lock(MSyn)
			{
				try
				{
					MHost.Open();
				}
				catch(Exception E)
				{
					Console.WriteLine(String.Format("CServiceHostWrapper.StartService() - Exception: {0} !",E.Message));
				}
			}
		}
//-------------------------------------------------------------------------------------------------------
		public void StopService()
		{
			lock(MSyn)
			{
				try
				{
					if (MHost.State==CommunicationState.Opened==true || MHost.State==CommunicationState.Opening==true)
						MHost.Close();
				}
				catch(Exception E)
				{
					Console.WriteLine(String.Format("CServiceHostWrapper.StopService() - Exception: {0} !",E.Message));
				}
			}
		}
//-------------------------------------------------------------------------------------------------------
		public void AddServiceEndpoint<TL>(Binding EndpointBinding, string Address)
		{
			lock(MSyn)
			{
				try
				{
					MHost.AddServiceEndpoint(typeof(TL),EndpointBinding,Address);
				}
				catch(Exception E)
				{
					Console.WriteLine(String.Format("CServiceHostWrapper.AddServiceEndpoint() - Exception: {0} !",E.Message));
				}
			}
		}
//-------------------------------------------------------------------------------------------------------
		public void AddTCPMEXEndpoint(string MEXEndpoindName)
		{
			lock(MSyn)
			{
				try
				{
					BindingElement			BE=new TcpTransportBindingElement();
					CustomBinding			CB=new CustomBinding(BE);
					ServiceMetadataBehavior SMB;

					if ((SMB=Host.Description.Behaviors.Find<ServiceMetadataBehavior>())==null)
					{
						SMB=new ServiceMetadataBehavior();
						Host.Description.Behaviors.Add(SMB);
					}

					MHost.AddServiceEndpoint(typeof(IMetadataExchange),CB,MEXEndpoindName);
				}
				catch(Exception E)
				{
					Console.WriteLine(String.Format("CServiceHostWrapper.AddServiceEndpoint() - Exception: {0} !",E.Message));
				}
			}
		}
//-------------------------------------------------------------------------------------------------------
	}
//-------------------------------------------------------------------------------------------------------
}
//-------------------------------------------------------------------------------------------------------